### PR TITLE
Table With Users

### PR DESCRIPTION
**Pull request by Aliaksandra-Leanovich**

### CODE REVIEW

File: spa-react/src/components/FormUser/FormUser.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  const { users } = useCreateUsers();
+  const { setUsersToDB } = useSetUsersToDb(users);
+  const { getUsers } = useGetUsersFromDB();
+
+  const onSubmit = () => {
+    const { name, surname } = getValues();
+    const user = {
+      name: name,
+      surname: surname,
+      id: "2",
```
Comment (rnavarych) [inline spa-react/src/components/FormUser/FormUser.tsx]: magic number

File: spa-react/src/hooks/use-generateUsers.hook.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  ];
+
+  const generateUsers = () => {
+    const rand_first = firstname[Math.floor(Math.random() * firstname.length)];
+    const rand_last = lastname[Math.floor(Math.random() * lastname.length)];
+
+    return {
+      name: rand_first,
+      surname: rand_last,
+      id: "1",
```
Comment (rnavarych) [inline spa-react/src/hooks/use-generateUsers.hook.ts]: magic number

File: spa-react/src/hooks/use-getUsers.hook.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+export const useGetUsersFromDB = () => {
+  const [usersFromDB, setUsersFromDB] = useState<IUser[]>();
+
+  const getUsers = () => {
+    const usersCollectionRef = collection(db, "usersDB");
+    getDocs(usersCollectionRef)
+      .then((response) => {
+        const users = response.docs.map((doc) => ({
```
Comment (rnavarych) [inline spa-react/src/hooks/use-getUsers.hook.ts]: Are you sure that docs exist?

File: spa-react/src/hooks/use-searchUser.hook.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+export const useSearchUser = () => {
+  const users = useAppSelector(getAllUsers);
+
+  const [filteredUsers, setFilteredUsers] = useState<IUser[]>([]);
+  const [searchValueName, setSearchValue] = useState<string>("");
+  const handleSearchName = (e: ChangeEvent<HTMLInputElement>) => {
+    setSearchValue(e.target.value);
+
+    setFilteredUsers(
+      users.users.users.filter((user) => {
```
Comment (rnavarych) [inline spa-react/src/hooks/use-searchUser.hook.ts]: users = undefined?

File: spa-react/src/hooks/use-searchUser.hook.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        return (
+          user.name.toLowerCase().includes(e.target.value.toLowerCase()) ||
```
Comment (rnavarych) [inline spa-react/src/hooks/use-searchUser.hook.ts:17]: undefined

File: spa-react/src/hooks/use-searchUser.hook.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+          user.surname.toLowerCase().includes(e.target.value.toLowerCase())
```
Comment (rnavarych) [inline spa-react/src/hooks/use-searchUser.hook.ts:18]: undefined

File: spa-react/src/pages/About.tsx [hunk 7]
```diff
@@ -1,6 +1,71 @@
...
+  }, [dispatch]);
+
+  useEffect(() => {
+    generateUsers();
+    setUsersToDB();
+    getUsers();
+  }, [users]);
+
+  useEffect(() => {
+    if (users.users.users.length && isLoading) {
```
Comment (rnavarych) [inline spa-react/src/pages/About.tsx]: undefined

File: spa-react/src/components/FormUser/FormUser.tsx [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    const { name, surname } = getValues();
+
+    const userNew = {
+      name: name,
+      surname: surname,
+      id: "2",
+      date: new Date().toLocaleString(),
+    };
+
+    dispatch(setNewUser(userNew));
```
Comment (rnavarych) [inline spa-react/src/components/FormUser/FormUser.tsx]: wrong data flow

File: spa-react/src/hooks/use-setUsers.hook.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+import { db } from "src/utils/firebase";
+
+interface IUsers {
+  users: IUser[];
+}
+
+export const useSetUsersToDb = (users: IUsers) => {
+  const setUsersToDB = async () => {
+    try {
+      const docRef = await setDoc(doc(db, "usersDB", "users"), users);
```
Comment (rnavarych) [inline spa-react/src/hooks/use-setUsers.hook.ts]: Each user should be written as new doc with unique id

File: spa-react/src/components/TableWithUsers/TableWithUsers.tsx [hunk 10]
```diff
@@ -0,0 +1,13 @@
+import { IProps } from "./types";
+
+export const TableWithUsers = ({ data }: IProps) => {
+  return (
+    <div>
+      {data.map((user, index) => (
```
Comment (rnavarych) [inline spa-react/src/components/TableWithUsers/TableWithUsers.tsx]: change to normal table

File: spa-react/src/store/slices/usersSlice.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  users: IUser[];
+}
+interface IInitialState {
+  users: {
+    users: IUser[];
+  };
+}
+const initialState: IInitialState = {
+  users: {
+    users: [],
```
Comment (rnavarych) [inline spa-react/src/store/slices/usersSlice.ts]: Why do you need to have users in users?

**Review by rnavarych — CHANGES_REQUESTED** [review]

File: spa-react/src/hooks/use-setUsers.hook.ts (hunk 9 above)
Comment (Aliaksandra-Leanovich) [reply spa-react/src/hooks/use-setUsers.hook.ts:12]: if there is no such doc, firestore will create it. I checked it, so we dont need to write condition for checking the existence of the doc

File: spa-react/src/hooks/use-getUsers.hook.ts (hunk 3 above)
Comment (Aliaksandra-Leanovich) [reply spa-react/src/hooks/use-getUsers.hook.ts]: if there is no such doc, firestore will create it. I checked it, so we dont need to write condition for checking the existence of the doc

File: spa-react/src/store/slices/usersSlice.ts (hunk 11 above)
Comment (Aliaksandra-Leanovich) [reply spa-react/src/store/slices/usersSlice.ts]: Firestore want to get data in object, thats why i have created object with key "users" and have putted array with 100 users there

File: spa-react/tsconfig.json [hunk 12]
```diff
@@ -1,6 +1,6 @@
 {
   "compilerOptions": {
-    "target": "es5",
+    // "target": "es5",
```
Comment (rnavarych) [inline spa-react/tsconfig.json]: Why do we need this comment?

File: spa-react/src/store/selectors/userSelector.ts [hunk 13]
```diff
@@ -1,3 +1,4 @@
 import { RootState } from "../store";
 
 export const getUserInfo = (state: RootState) => state.user;
+export const getAllUsers = (state: RootState) => state.users.users;
```
Comment (rnavarych) [inline spa-react/src/store/selectors/userSelector.ts:4]: why do you use this structure of Users entity? users.users

File: spa-react/src/pages/About.tsx [hunk 14]
```diff
@@ -1,6 +1,89 @@
...
+        padding: "30px 0",
+        width: "100%",
+        justifyContent: "center",
+        alignItems: "center",
+      }}
+    >
+      <SearchInput
+        searchValue={searchValueName}
+        handleSearch={handleSearchName}
+        placeholder="Search by name/surname"
```
Comment (rnavarych) [inline spa-react/src/pages/About.tsx]: Why do you use hardcode again!?

File: spa-react/src/pages/About.tsx [hunk 15]
```diff
@@ -1,6 +1,89 @@
...
+      <ModalForm show={show} showModal={showModal} />
+      {isLoading ? (
+        <div>Loading....</div>
+      ) : (
+        <>
+          <div>
+            {filteredUsers?.length > 0 ? (
+              <Table data={filteredUsers} />
+            ) : (
+              <p>Oooops 🙈</p>
```
Comment (rnavarych) [inline spa-react/src/pages/About.tsx]: hardcode

File: spa-react/src/pages/About.tsx [hunk 16]
```diff
@@ -1,6 +1,89 @@
...
+      }}
+    >
+      <SearchInput
+        searchValue={searchValueName}
+        handleSearch={handleSearchName}
+        placeholder="Search by name/surname"
+      />
+      <ModalForm show={show} showModal={showModal} />
+      {isLoading ? (
+        <div>Loading....</div>
```
Comment (rnavarych) [inline spa-react/src/pages/About.tsx]: hardcode

File: spa-react/src/pages/About.tsx [hunk 17]
```diff
@@ -1,6 +1,89 @@
...
+  };
+
+  useEffect(() => {
+    dispatch(setInitialUsers(usersFromDB));
+    setFilteredUsers(usersFromDB);
+    console.log(users);
+  }, [dispatch, usersFromDB]);
+
+  useEffect(() => {
+    generateUsers();
```
Comment (rnavarych) [inline spa-react/src/pages/About.tsx]: this is wrong flow. Should be one data direction flow

File: spa-react/src/hooks/use-getUsers.hook.ts (hunk 3 above)
Comment (rnavarych) [reply spa-react/src/hooks/use-getUsers.hook.ts]: If you don't have any docs you will have have a crash in doc.data()

File: spa-react/src/hooks/use-generateUsers.hook.ts [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+      name: rand_first,
+      surname: rand_last,
+      id: uuid(),
+      date: new Date().toLocaleString().slice(0, 10).replace(/-/g, "/"),
+      sum: sum,
+    };
+  };
+
+  const generatedUsers = [
+    ...new Set(Array.from({ length: 100 }, () => generateUsers())),
```
Comment (rnavarych) [inline spa-react/src/hooks/use-generateUsers.hook.ts]: magic number

File: spa-react/src/hooks/use-generateUsers.hook.ts [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+    const rand_first = firstname[Math.floor(Math.random() * firstname.length)];
+    const rand_last = lastname[Math.floor(Math.random() * lastname.length)];
+
+    let sum = calculateCodeSum(rand_last);
+
+    return {
+      name: rand_first,
+      surname: rand_last,
+      id: uuid(),
+      date: new Date().toLocaleString().slice(0, 10).replace(/-/g, "/"),
```
Comment (rnavarych) [inline spa-react/src/hooks/use-generateUsers.hook.ts]: move to utils

File: spa-react/src/components/Table/Table.tsx [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+      setFiltered(data);
+    }
+  };
+
+  return (
+    <TableContainer component={Paper}>
+      <MuiTable sx={{ width: "800px" }} aria-label="simple table">
+        <TableHead>
+          <TableRow>
+            <TableCell sx={{ fontWeight: "600" }}>Name</TableCell>
```
Comment (rnavarych) [inline spa-react/src/components/Table/Table.tsx]: hardcode again! You should review all your code!

File: spa-react/src/components/FormUser/FormUser.tsx [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+          )}
+          rules={{
+            required: false,
+          }}
+        />
+        <Controller
+          name="surname"
+          control={control}
+          render={({ field: { onChange } }) => (
+            <Input type="surname" placeholder="Surname" onChange={onChange} />
```
Comment (rnavarych) [inline spa-react/src/components/FormUser/FormUser.tsx:53]: -//-

File: spa-react/src/components/FormUser/FormUser.tsx [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+      id: uuid(),
+      date: new Date().toLocaleString(),
+      sum: sum,
+    };
+    setUsersToDB(userNew.id, userNew);
+
+    showModal(!show);
+    reset();
+
+    getUsers();
```
Comment (rnavarych) [inline spa-react/src/components/FormUser/FormUser.tsx]: wrong flow

**Review by rnavarych — CHANGES_REQUESTED** [review]

**Pull request merged**